File: cpp/egrabber-sample-programs/python/240-user-memory.py
"""Grab into user allocated buffer"""

from egrabber import *

gentl = EGenTL()
grabber = EGrabber(gentl)
payload_size = grabber.get_payload_size()
user_buffer = bytearray(payload_size)
grabber.announce_and_queue(UserMemory(user_buffer))
grabber.start(1)
with Buffer(grabber) as buffer:
    base = buffer.get_info(BUFFER_INFO_BASE, INFO_DATATYPE_PTR)
    data_size = buffer.get_info(BUFFER_INFO_DATA_SIZE, INFO_DATATYPE_SIZET)
    print('User memory address: {}, buffer data size: {}'.format(base, data_size))
grabber.stop()
grabber.realloc_buffers(0)

File: cpp/egrabber-sample-programs/cpp/egrabber-snippets/samples/240-user-memory.cpp
#include "../tools/tools.h"
#include <EGrabber.h>

using namespace Euresys;

static void sample() {
    EGenTL genTL;
    EGrabber<CallbackOnDemand> myGrabber(genTL);
    size_t payLoadSize = myGrabber.getPayloadSize();
    char* userBuffer = new char[payLoadSize];
    // Announce and queue user buffer
    myGrabber.announceAndQueue(UserMemory(userBuffer, payLoadSize));
    myGrabber.start(1);
    ScopedBuffer buffer(myGrabber);
    Tools::log("User memory address: " + 
               Tools::toHexString(buffer.getInfo<void *>(gc::BUFFER_INFO_BASE)) + 
               ", buffer data size: " + 
               Tools::toString(buffer.getInfo<size_t>(gc::BUFFER_INFO_DATA_SIZE)));
    myGrabber.stop(); // Make sure the data stream is stopped before revoking the buffer
    myGrabber.reallocBuffers(0); // Revoke announced user buffer before deleting it
    delete[] userBuffer;
}

static Tools::Sample addSample(__FILE__, sample, "Grab into user allocated buffer");
